### PR TITLE
Replace Deprecated Exceptions

### PR DESCRIPTION
**Pull request by ravjotbrar**
### Summary

Replaced dbt-core's deprecated exceptions with new exceptions

### Description

Replaced the following old exceptions with new exceptions

Old | New
-- | --
RuntimeException | DbtRuntimeError
DatabaseException | DbtDatabaseError
raise_compiler_error() | CompilationError

### Test Results

Successfully ran all functional tests

### Changelog

-   [x] Added a summary of what this PR accomplishes to CHANGELOG.md

### Related Issue

#117

### CODE REVIEW

**Comment by ravjotbrar** [comment]
Ran the unit tests and it seems like the FailedToConnectException was removed entirely. Updated to new exception to get all unit tests passing.

File: tests/unit/test_connection.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ def test_connection_retry(
         )
 
         # Act
-        with pytest.raises(FailedToConnectException):
+        with pytest.raises(FailedToConnectError):
```
Comment (ArgusLi) [inline tests/unit/test_connection.py:44]: Should this be the dbtRuntimeError? Based on [this documentation](https://docs.getdbt.com/guides/best-practices/debugging-errors), failing to connect to the Dremio instance should result in a dbtRuntimeError right? Doe we have to change what error is raised when connecting?

File: dbt/include/dremio/macros/utils/array_append.sql [hunk 2]
```diff
@@ -13,5 +13,5 @@ See the License for the specific language governing permissions and
 limitations under the License.*/
 
 {% macro dremio__array_append(array, new_element) -%}
-    {% do exceptions.raise_compiler_error("Array Append is not currently supported by Dremio.") %}
+    {% do exceptions.CompilationError("Array Append is not currently supported by Dremio.") %}
```
[the remark this answers is not in the document]
Comment (ArgusLi) [reply dbt/include/dremio/macros/utils/array_append.sql:16]: In addition should the array errors raised actually be dbtDatabaseErrors instead?

File: dbt/include/dremio/macros/utils/array_concat.sql [hunk 3]
```diff
@@ -13,5 +13,5 @@ See the License for the specific language governing permissions and
 limitations under the License.*/
 
 {% macro dremio__array_concat(array_1, array_2) -%}
-    {% do exceptions.raise_compiler_error("Array Concat is not currently supported by Dremio.") %}
+    {% do exceptions.CompilationError("Array Concat is not currently supported by Dremio.") %}
```
Comment (ArgusLi) [inline dbt/include/dremio/macros/utils/array_concat.sql:16]: This is a nit: should we have `CompilationError(msg="Array Concat..")` instead?

File: dbt/include/dremio/macros/utils/array_construct.sql [hunk 4]
```diff
@@ -16,6 +16,6 @@ limitations under the License.*/
     {% if inputs|length > 0 %}
     array[ {{ inputs|join(' , ') }} ]
     {% else %}
-    {% do exceptions.raise_compiler_error("Constructing empty arrays is not currently supported by Dremio.") %}
+    {% do exceptions.CompilationError("Constructing empty arrays is not currently supported by Dremio.") %}
```
Comment (ArgusLi) [inline dbt/include/dremio/macros/utils/array_construct.sql:19]: This is a nit: should we have `CompilationError(msg="Constructing..")` instead?

File: tests/functional/adapter/basic/test_adapter_methods.py [hunk 5]
```diff
@@ -26,7 +26,7 @@
     {%- do adapter.drop_schema(upstream) -%}
     {% set existing = adapter.get_relation(upstream.database, upstream.schema, upstream.identifier) %}
     {% if existing is not defined %}
-        {% do exceptions.raise_compiler_error('expected ' ~ ' to not exist, but it did') %}
+        {% do exceptions.CompilationError('expected ' ~ ' to not exist, but it did') %}
```
Comment (ArgusLi) [inline tests/functional/adapter/basic/test_adapter_methods.py:29]: This is a nit: should we have `CompilationError(msg="expected...")` instead?

File: dbt/include/dremio/macros/adapters/apply_grants.sql [hunk 6]
```diff
@@ -28,7 +28,7 @@ limitations under the License.*/
     {%- elif target.software_host and not target.cloud_host -%}
         {%- set privileges_table = 'sys.privileges' -%}
     {%- else -%}
-         {% do exceptions.raise_compiler_error("Invalid profile configuration: please only specify one of cloud_host or software_host in profiles.yml") %}
+         {% do exceptions.CompilationError("Invalid profile configuration: please only specify one of cloud_host or software_host in profiles.yml") %}
```
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply dbt/include/dremio/macros/adapters/apply_grants.sql:31]: Good suggestion, but I don't completely see the value in adding msg before the message.

File: dbt/include/dremio/macros/utils/array_append.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (ravjotbrar) [reply dbt/include/dremio/macros/utils/array_append.sql:16]: According to the documentation you linked, it seems like any errors raised in JInja macros should be CompilationErrors no?

File: tests/unit/test_connection.py (hunk 1 above)
Comment (ravjotbrar) [reply tests/unit/test_connection.py:44]: I'll change to dbtRuntimeError and test. Thanks for the suggestion.

File: dbt/include/dremio/macros/utils/array_append.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArgusLi) [reply dbt/include/dremio/macros/utils/array_append.sql:16]: The thing is that the Jinja snippets and yml files are valid, it's just that it's technically a database issue in that we don't support it? I'm not quite sure which one it applies to to be honest.

[the remark this answers is not in the document]
Comment (ArgusLi) [reply dbt/include/dremio/macros/utils/array_append.sql:16]: Let's leave it as a compiler error for now. I guess it applies to that more considering we are not accepting that macro at all in the dbt-dremio stage.

**Review by ArgusLi — APPROVED** [review]
Looks good to me!

**Pull request merged**